docs/fundamentals/gateway/custom-gateway.md

```diff
@@ -0,0 +1,376 @@
+(custom-gateway)=
+# Customize the Gateway
+Jina Gateways are customizable, and you can implement them in much the same way as an Executor.
+With Custom Gateways, Jina gives power back to the users, by allowing them to implement any server, protocol and 
+interface at the gateway level. This means you have more freedom to:
+* Define and expose your own API Gateway interface to clients. You can define your JSON schema or protos,...
+* Choose your favourite server framework.
+* Choose the protocol used to serve your app.
+
+The next sections will detail the steps to implement and use a custom Gateway.
+
+## Implementing the custom Gateway
+Just like for Executors, you can implement a custom Gateway by inheriting from a base gateway class.
+Jina will instantiate your implemented class, inject runtime arguments and user-defined arguments into it, 
+run it, orchestrate it, and send it health-checks.
+
+There are two Gateway base classes to allow implementing a custom gateway:
+* {class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway`: Use this abstract class to implement a custom gateway using FastAPI.
+* {class}`~jina.Gateway`: Use this abstract class to implement a custom gateway of any type.
+
+### Using {class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway`:
+{class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway` offers a simple API to implement custom 
+gateways but is restricted to FastAPI apps.
+
+In order to implement a custom gateway using {class}`~jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway`, 
+simply implement the {meth}`~jina.jina.serve.runtimes.gateway.http.fastapi.FastAPIBaseGateway.app` property:
```

Comment:
To me it is still not so clear what I would have to do here. I implement the `app` property, but what do I have to put there? Do I have to convert incoming requests to docarray? do I have to use the GatewayStreamer? Is some of this done automatically for me? In short, what is my job as the implementer here?